picard/config.py

```diff
@@ -129,6 +131,62 @@ def value(self, name, option_type, default=None):
         return default
 
 
+class SettingConfigSection(ConfigSection):
+    """Custom subclass to automatically accommodate saving and retrieving values based on user profile settings.
+    """
+    def __init__(self, config, name):
+        super().__init__(config, name)
+        self.__qt_config = config
+        self.__name = name
+        self.__prefix = self.__name + '/'
+        self._memoization = defaultdict(Memovar)
+        ListOption.add_if_missing(name, 'user_profiles', [])
+        Option.add_if_missing(name, 'user_profile_settings', {})
+
+    def _get_active_profile_ids(self):
+        profiles = self['user_profiles']
+        if profiles is None:
+            profiles = []
+        for profile in profiles:
+            if profile['active']:
+                yield profile["id"]
+
+    def _get_profile_settings(self, id):
+        profile_settings = self["user_profile_settings"][id]
+        if profile_settings is None:
+            log.error("Unable to find settings for user profile '%s'", id)
+            return {}
+        return profile_settings
+
+    def __getitem__(self, name):
+        # Don't process settings that are not profile-specific
+        if name in UserProfileGroups.get_all_settings_list():
+            for id in self._get_active_profile_ids():
+                profile_settings = self._get_profile_settings(id)
+                if name in profile_settings and profile_settings[name] is not None:
+                    return profile_settings[name]
+        opt = Option.get(self.__name, name)
+        if opt is None:
+            return None
+        return self.value(name, opt, opt.default)
+
+    def __setitem__(self, name, value):
+        # Don't process settings that are not profile-specific
+        if name in UserProfileGroups.get_all_settings_list():
+            for id in self._get_active_profile_ids():
+                settings = self._get_profile_settings(id)
```

Comment:
A generator could replace those 2 lines:

```python
def _get_active_profile_settings_with_id(self):
    for id in self._get_active_profile_ids():
        yield id, self._get_profile_settings(id)
```

```python
for id, profile_settings in self._get_active_profile_settings_with_id():
    ....
                self["user_profile_settings"][id] = settings
```

Reply:
I'll actually use this same generator for the `_get_active_profile_settings()` and just ignore the `id` part of the tuple returned.  That should avoid some code duplication.